_posts/2024-12-04-enable-pluggable-storage-in-opensearch-vectordb.md

```diff
@@ -0,0 +1,255 @@
+
+```json
+POST knn-index/_search
+{
+  "query": {
+    "knn": {
+      "my_vector": {
+        "vector": [2, 3],
+        "k": 2
+      }
+    }
+  }
+}
+```
+
+The response returns vectors closest to the query vector:
+
+```json
+{
+  "took": 16,
+  "timed_out": false,
+  "_shards": {
+    "total": 1,
+    "successful": 1,
+    "skipped": 0,
+    "failed": 0
+  },
+  "hits": {
+    "total": {
+      "value": 2,
+      "relation": "eq"
+    },
+    "max_score": 0.6666667,
+    "hits": [
+      {
+        "_index": "knn-index",
+        "_id": "1",
+        "_score": 0.6666667,
+        "_source": {
+          "my_vector": [
+            1.5,
+            2.5
+          ],
+          "price": 12.2
+        }
+      },
+      {
+        "_index": "knn-index",
+        "_id": "2",
+        "_score": 0.6666667,
+        "_source": {
+          "my_vector": [
+            2.5,
+            3.5
+          ],
+          "price": 7.1
+        }
+      }
+    ]
+  }
+}
+```
+
+### Step 4: Take a snapshot
+
+Take a snapshot of the index. For detailed steps, see [Take and restore snapshots](https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/snapshots/snapshot-restore/). After taking the snapshot, delete the `knn-index` so it is no longer available locally.
+
+### Step 5: Create a searchable snapshot index from the snapshot
+
+Use the following request to restore the original index from the snapshot as a searchable snapshot:
+
+```json
+POST _snapshot/<SNAPSHOT_REPO>/<SNAPSHOT_NAME>/_restore
+{
+  "storage_type": "remote_snapshot",
+  "indices": "knn-index"
+}
+```
+
+To verify the successful creation of the searchable snapshot index, use the following request:
+
+```json
+GET /_cat/indices
+```
+For more information, see [Create a searchable snapshot index](https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/snapshots/searchable_snapshot/#create-a-searchable-snapshot-index).
+
+### Step 6: Run a vector search query
+
+Run a vector search query on the searchable snapshot index:
+
+```json
+POST knn-index/_search
+{
+  "query": {
+    "knn": {
+      "my_vector": {
+        "vector": [2, 3],
+        "k": 2
+      }
+    }
+  }
+}
+```
+
+The query returns the same results as the query on the local index in [Step 3](#step-3-query-the-local-index).
+
+## Conclusion
+
+By introducing an abstract I/O layer that uses Lucene's **Directory** abstraction, we eliminated the native engines' dependency on file-based APIs that limited storage to local file systems. This change allows the vector engine to read graph data structures from any storage system supported by OpenSearch's **Directory** implementation. Our extensive benchmark testing confirmed that this change maintained the original file API-based approach's search performance. Notably, we observed no regression in search times after the graphs were loaded into memory (graph loading is a one-time operation for a properly scaled cluster).
+
```

Comment:
Line 247, first sentence: I assumed that we didn't need both "abstract" and "abstraction" in the same sentence.